cplusplus/gen-operators.py

```diff
@@ -1,212 +1,257 @@
-#!/usr/bin/python
+#!/usr/bin/env python
 
-# walk vips and generate member definitions for all operators
+# This file generates the member definitions and declarations for all vips operators.
+# It's in Python, since we use the whole of FFI.
 
-# sample member definition:
+# Regenerate the files with something like:
+#
+#   cd cplusplus
+#   python gen-operators.py
 
-# VImage VImage::invert( VOption *options )
-# {
-# 	VImage out;
-# 
-# 	call( "invert", 
-# 		(options ? options : VImage::option())-> 
-# 			set( "in", *this )->
-# 			set( "out", &out ) );
-# 
-# 	return( out );
-# }
+# this needs pyvips
+#
+#   pip install --user pyvips
 
-import sys
-import re
+# Sample member declaration:
+# VImage invert(VOption *options = 0) const;
 
-import logging
-#logging.basicConfig(level = logging.DEBUG)
+# Sample member definition:
+# VImage VImage::invert( VOption *options ) const
+# {
+#     VImage out;
+#
+#     call( "invert",
+#         (options ? options : VImage::option())->
+#             set( "in", *this )->
+#             set( "out", &out ) );
+#
+#     return( out );
+# }
 
-import gi
-gi.require_version('Vips', '8.0')
-from gi.repository import Vips, GObject
+import datetime
 
-vips_type_image = GObject.GType.from_name("VipsImage")
-vips_type_operation = GObject.GType.from_name("VipsOperation")
-param_enum = GObject.GType.from_name("GParamEnum")
+from pyvips import Image, Operation, GValue, Error, \
+    ffi, gobject_lib, type_map, type_from_name, nickname_find, type_name
 
 # turn a GType into a C++ type
 gtype_to_cpp = {
-    "VipsImage" : "VImage",
-    "gint" : "int",
-    "gdouble" : "double",
-    "gboolean" : "bool",
-    "gchararray" : "char *",
-    "VipsArrayInt" : "std::vector<int>",
-    "VipsArrayDouble" : "std::vector<double>",
-    "VipsArrayImage" : "std::vector<VImage>",
-    "VipsBlob" : "VipsBlob *"
+    GValue.gbool_type: 'bool',
+    GValue.gint_type: 'int',
+    GValue.gdouble_type: 'double',
+    GValue.gstr_type: 'char *',
+    GValue.refstr_type: 'char *',
+    GValue.gflags_type: 'int',
+    GValue.gobject_type: 'char *',
```

Comment:
This is curious. Do we need one for gobject? It seems to work OK without this line.

Reply:
Removed (leftover from: https://github.com/libvips/php-vips/blob/master/examples/generate_phpdoc.py#L29).